### ISSUE TITLE
Question on convergence

### ISSUE THREAD

**Issue by ppbrown** [konstmish/prodigy#18]
Disclaimer: very new at all this. I'm coming from the perspective of an inference user, when on a good model using a nice prompt and sampler, there tends to be a number of steps past which, it will always converge to a nice stable image for a given seed.

I think I've read that there is some similar effect in (SDXL specifically) model training, where if you are doing things right, you will get "convergence" for training the resulting model.
I take it to mean that I would see similar things if I viewed the per-epoch sample images; they would converge to something decent over the course of the epochs.

But... I'm not seeing that happening.
For example, if I run a 100 epoch training over 80 images, I will see something reasonable congeal around maybe epoch 25.... 
and then it gets mushy for a while.. and then things come back into focus around epoch 80.

I tried turning on "safeguard_warmup" and "bias correction".. but the overall effect of those combined, seemed to be to just stretch out the training cycle. Now the things that happened at 80, happen at 2x80=160 epochs. (literally almost the same images)

Are my expectations off? Is it reasonable to believe that there ARE settings that will not only have a convergence, but will converge on something sane looking, given a good input dataset?

Im using onetrainer with the following settings at present:

```
scheduler: constant
learning rate: 1
warmup steps: 200
learning rate cycles: 1
epochs 100
batchsize: 1
gradient accumulation steps: 1
ema: GPU - ema decay 0.999 - update step interval: 5

beta1: 0.9
beta2: 0.999
d0: 3e-06
d_coeficient: 1
decouple: true
eps: 1e-08
relative step: false
safeguard: (true/false)
bias_correction: (true/false)

weight decay: 0.01

```

**Comment by riffmaster-2001** [comment]
have you tried COSINE for onetrainer? the paper also says use Cosine Annealing and onetrainer's COSINE is equivalent (reference: https://github.com/Nerogar/OneTrainer/issues/214) 

Also try increasing your  d_coeficient to 2.0 or 3.0 which will let it jump to higher LRs as it searches for the right LR. Watch tensorboard to see where it's jumping to

**Comment by ppbrown** [comment]
I'm confused. I though the "COSINE" stuff, actually reset the LR and made for MORE varience.
Which intuitively makes me think its for the opposite ?

"Cosine Annealing is a type of learning rate schedule that has the effect of starting with a large learning rate that is relatively rapidly decreased to a minimum value before being increased rapidly again"

**Comment by konstmish** [comment]
Sorry for the late response, I realize you might not be interested in it anymore but posting since others might find this issue.

In theory, the images might be changing over the course of training many times without converging to anything specific. This is because as long as the model is training, its parameters keep changing, this effect should be also visible with Adam and its variants (Prodigy itself is a variant of Adam with on-the-fly estimation of the learning rate). As mentioned, cosine annealing with `safeguard_warmup` might forces the optimizer to slow down eventually, though the exact images you'll see depend on other settings, especially on the length of the training.

I'd also like to add that we won't really know why images may get mushy or what makes images converge to any other distribution. What we should expect in general is that the loss should go down with enough trainable parameters in the model, but how this will affect things down the line is usually unclear.

**Comment by ppbrown** [comment]
Thanks for the reply.
right now I'm playing with dadapt-lion, since lower memory requirements.
With that one I'm seeing it increasing for a while, then plateau... then increasing again eventually. 
For that one, I read that it is actually developed with cosine scheduler in mind, so it almost makes sense.
But I didnt think prodigy was like that too?
Both of them seem to do super well, with shorter runs, eg: 10k steps or less. But if you get up to 20k, feels like it overtrains.

**Comment by Xephier102** [comment]
> I'd also like to add that we won't really know why images may get mushy or what makes images converge to any other distribution. What we should expect in general is that the loss should go down with enough trainable parameters in the model, but how this will affect things down the line is usually unclear.

Just wanted to add to what you said, for the newbies, focus on 'smooth loss', not 'loss'. Cuz regular loss is only loss per image. So, if an epoch/training run ends at 5.0 loss, 0.001 smooth loss, then you've got a very good model despite that one image doing terribly bad. 

And to the owner, if you happen to get back to this in the next day or few, I recently tried training prodigy with EMA, Never tried EMA before, but I put update steps to 251 (same number of images in my dataset), and it didn't train the ema at all. Made for some nice 'no ema' samples, but a useless lora in the end. Not sure if it did this because EMA update step was too high, or cuz I'm a moron and messed somethin else up. Conversely, I also enabled training Unet, which is something else I also never do. Weirdly it seems to be taking less Vram with those options on.. I'm using Onetrainer GUI for this. and Cosine for scheduler. Oh, and yes, Unet, and LR of text encoder embeddings and base-LR all set to 1.0. 

It took me a while to get around to using prodigy, cuz ChatGPT absolutely insisted to never use something as high as 1.0 as an LR, but yea, when using less, nothing.. lol.. I got a great lora trained the last time, but then forgot to save the settings.. So, I'm kickin myself for that. And I had tensorboard turned off to save resources, so there's like, no record of them either.

**Comment by konstmish** [comment]
Hey Xepheir102, thanks for chiming in. I am not sure I understand why the lora was useless in the end. Are you saying that EMA parameters are not good but the final parameters are? If that's so, why can't you use the final parameters for the lora?

I'm not sure why turning on training Unet could make it more memory efficient, that sounds quite strange, and normally nothing like that should happen.

I understand why ChatGPT suggested not using a learning rate of 1.0, it wouldn't make sense for normal Adam. We do say in the Readme that "We recommend using lr=1. (default) for all networks."

**Comment by Xephier102** [comment]
> Hey Xepheir102, thanks for chiming in. I am not sure I understand why the lora was useless in the end. Are you saying that EMA parameters are not good but the final parameters are? If that's so, why can't you use the final parameters for the lora?
> 
> I'm not sure why turning on training Unet could make it more memory efficient, that sounds quite strange, and normally nothing like that should happen.
> 
> I understand why ChatGPT suggested not using a learning rate of 1.0, it wouldn't make sense for normal Adam. We do say in the Readme that "We recommend using lr=1. (default) for all networks."

Hello, I actually got your reply right away, but tbh, I was a bit embarrassed at the time because as it turns out, I was mistaken about at least one of the things I claimed. I went over all my prior training presets, and it seems I'd actually trained Unet on every run. I just never actually paid attention to it before prodigy, as I left it's LR on 0, or empty, or whatever the default was. 

So, yes, Unet wasn't a part of the issue at all. But it still did seem to use less Vram with EMA turned on, which I guarantee I had not attempted doing on prior runs, so my negative experience with it thus far, is most likely due to some flaw that I've made due to my lack of understanding in how it works, or what settings to use with it. 

As for why the Lora was useless; the thing is, it had trained EMA very badly(as in, basically not at all, end samples looked same as first sample), likely due to setting the update steps to the same as the number of images in the dataset. The non-ema (AKA: 'no ema' as it had labelled the sample folder), was trained quite well, apparently, however, when using the lora in ComfyUI, it was clearly running inference using the EMA weights and not the non-ema weights, hence, useless lora. 

Honestly, the whole idea of having two sets of weights on one lora, confuses the heck outta me. I mean, I can roughly understand it for training, but to have them both on the end-product(Lora) itself, I don't get.. 

I have tried training since then using lower ema update steps(one run on 1, and another on 5, which was the initial default). But, I could tell right away that the outcome would not be as good for the end-product as it was for the non-ema weights on the first lora; as the convergence was 0.03-0.04(smooth loss) greater than the first run, and, this is a bit random, but the first couple sample images for non ema in the first run came out as purely 1 color, white then green for the second(I think it was). 

Like as if someone had opened them in paint and used the paint bucket tool on the entire image, but after that, the images became good. But on the run with the ema update steps turned down, those same images, were different colors, as with the first run, but instead of being purely one color, they looked like 'snow' on an old television(CRT TV's when they had no signal). In other words, very 'noisy'. And they didn't seem to become cohesive in the followup steps either. 

I did try a run with EMA turned off, but wasn't able to replicate the same level of convergence as the first run, which was 0.06 or so in the smooth loss.  

"it wouldn't make sense for normal Adam."

Or for any other optimizers that I'm currently aware of.

"We do say in the Readme that "We recommend using lr=1. (default) for all networks.""

Yes, I am aware, I ended up going over that more recently, hence setting everything to 1.0. Though I don't believe I was able to find any notes regarding EMA, or if EMA even acts any differently with prodigy than with other optimizers.
